diff --git a/pkg/auth/providers/ldap/ldap_actions.go b/pkg/auth/providers/ldap/ldap_actions.go
index 79ca7aae4..ee50a4655 100644
--- a/pkg/auth/providers/ldap/ldap_actions.go
+++ b/pkg/auth/providers/ldap/ldap_actions.go
@@ -1,13 +1,16 @@
 package ldap
 
 import (
+	"encoding/json"
 	"fmt"
+	"net/http"
 	"strings"
 
 	ldapv3 "github.com/go-ldap/ldap/v3"
 	"github.com/rancher/norman/api/handler"
 	"github.com/rancher/norman/httperror"
 	"github.com/rancher/norman/types"
+	"github.com/rancher/norman/types/convert"
 	v3 "github.com/rancher/rancher/pkg/apis/management.cattle.io/v3"
 	"github.com/rancher/rancher/pkg/auth/providers/common"
 	"github.com/rancher/rancher/pkg/auth/providers/common/ldap"
@@ -21,6 +24,7 @@
 func (p *ldapProvider) formatter(apiContext *types.APIContext, resource *types.RawResource) {
 	common.AddCommonActions(apiContext, resource)
 	resource.AddAction(apiContext, "testAndApply")
+	resource.AddAction(apiContext, "search")
 }
 
 func (p *ldapProvider) actionHandler(actionName string, action *types.Action, request *types.APIContext) error {
@@ -35,6 +39,8 @@ func (p *ldapProvider) actionHandler(actionName string, action *types.Action, re
 	switch actionName {
 	case "testAndApply":
 		return p.testAndApply(request)
+	case "search":
+		return p.search(request)
 	default:
 		return httperror.NewAPIError(httperror.ActionNotAvailable, "")
 	}
@@ -172,6 +178,48 @@ func (p *ldapProvider) testAndApply(request *types.APIContext) error {
 	return p.tokenMGR.CreateTokenAndSetCookie(user.Name, userPrincipal, groupPrincipals, "", 0, "Token via LDAP Configuration", request)
 }
 
+type searchPrincipalsInput struct {
+	Name          string `json:"name"`
+	PrincipalType string `json:"principalType"`
+	Page          int    `json:"page"`
+	PageSize      int    `json:"pageSize"`
+}
+
+func (p *ldapProvider) search(request *types.APIContext) error {
+	var input searchPrincipalsInput
+	if err := json.NewDecoder(request.Request.Body).Decode(&input); err != nil {
+		return httperror.NewAPIError(httperror.InvalidBodyContent, fmt.Sprintf("Failed to parse body: %v", err))
+	}
+
+	config, caPool, err := p.getLDAPConfig(p.authConfigs.ObjectClient().UnstructuredClient())
+	if err != nil {
+		return err
+	}
+
+	lConn, err := ldap.Connect(config, caPool)
+	if err != nil {
+		return err
+	}
+	defer lConn.Close()
+
+	principals, err := p.searchPrincipals(input.Name, input.PrincipalType, config, lConn, input.Page, input.PageSize)
+	if err != nil {
+		return err
+	}
+
+	var result []map[string]any
+	for _, principal := range principals {
+		data, err := convert.EncodeToMap(principal)
+		if err != nil {
+			return err
+		}
+		result = append(result, data)
+	}
+
+	request.WriteResponse(http.StatusOK, result)
+	return nil
+}
+
 func (p *ldapProvider) saveLDAPConfig(config *v3.LdapConfig) error {
 	storedConfig, _, err := p.getLDAPConfig(p.authConfigs.ObjectClient().UnstructuredClient())
 	if err != nil {
diff --git a/pkg/auth/providers/ldap/ldap_client.go b/pkg/auth/providers/ldap/ldap_client.go
index ed6caea6c..616f5ca9b 100644
--- a/pkg/auth/providers/ldap/ldap_client.go
+++ b/pkg/auth/providers/ldap/ldap_client.go
@@ -168,7 +168,7 @@ func (p *ldapProvider) getPrincipalsFromSearchResult(result *ldapv3.SearchResult
 		query = fmt.Sprintf("(&%s%s)", filter, query)
 		// Pulling user's groups
 		logrus.Debugf("Ldap: Query for pulling user's groups: %s", query)
-		userMemberGroupPrincipals, err := p.searchLdap(query, groupScope, config, lConn)
+		userMemberGroupPrincipals, err := p.searchLdap(query, groupScope, config, lConn, 0, 0)
 		groupPrincipals = append(groupPrincipals, userMemberGroupPrincipals...)
 		if err != nil {
 			return userPrincipal, groupPrincipals, err
@@ -193,7 +193,7 @@ func (p *ldapProvider) getPrincipalsFromSearchResult(result *ldapv3.SearchResult
 			ObjectClass,
 			ldap.SanitizeAttr(config.GroupObjectClass),
 		)
-		newGroupPrincipals, err := p.searchLdap(query, groupScope, config, lConn)
+		newGroupPrincipals, err := p.searchLdap(query, groupScope, config, lConn, 0, 0)
 		if err != nil {
 			return userPrincipal, groupPrincipals, err
 		}
@@ -216,7 +216,7 @@ func (p *ldapProvider) getPrincipalsFromSearchResult(result *ldapv3.SearchResult
 			ObjectClass,
 			ldap.SanitizeAttr(config.GroupObjectClass),
 		)
-		groupPrincipals, err = p.searchLdap(query, groupScope, config, lConn)
+		groupPrincipals, err = p.searchLdap(query, groupScope, config, lConn, 0, 0)
 		if err != nil {
 			return userPrincipal, groupPrincipals, err
 		}
@@ -364,11 +364,11 @@ func (p *ldapProvider) getPrincipal(distinguishedName string, scope string, conf
 	return principal, nil
 }
 
-func (p *ldapProvider) searchPrincipals(name, principalType string, config *v3.LdapConfig, lConn ldapv3.Client) ([]v3.Principal, error) {
+func (p *ldapProvider) searchPrincipals(name, principalType string, config *v3.LdapConfig, lConn ldapv3.Client, page, pageSize int) ([]v3.Principal, error) {
 	var principals []v3.Principal
 
 	if principalType == "" || principalType == "user" {
-		userPrincipals, err := p.searchUser(name, config, lConn)
+		userPrincipals, err := p.searchUser(name, config, lConn, page, pageSize)
 		if err != nil {
 			return nil, err
 		}
@@ -376,7 +376,7 @@ func (p *ldapProvider) searchPrincipals(name, principalType string, config *v3.L
 	}
 
 	if principalType == "" || principalType == "group" {
-		groupPrincipals, err := p.searchGroup(name, config, lConn)
+		groupPrincipals, err := p.searchGroup(name, config, lConn, page, pageSize)
 		if err != nil {
 			return nil, err
 		}
@@ -386,7 +386,7 @@ func (p *ldapProvider) searchPrincipals(name, principalType string, config *v3.L
 	return principals, nil
 }
 
-func (p *ldapProvider) searchUser(name string, config *v3.LdapConfig, lConn ldapv3.Client) ([]v3.Principal, error) {
+func (p *ldapProvider) searchUser(name string, config *v3.LdapConfig, lConn ldapv3.Client, page, pageSize int) ([]v3.Principal, error) {
 	if config.UserSearchFilter != "" {
 		// Make sure user search filter contains a valid LDAP query expression
 		// before interpolating it into the search filter.
@@ -410,10 +410,10 @@ func (p *ldapProvider) searchUser(name string, config *v3.LdapConfig, lConn ldap
 	// and is expected to follow ldap syntax and enclosed in parentheses.
 	query += srchAttrs + ")" + config.UserSearchFilter + ")"
 	logrus.Debugf("%s searchUser query: %s", p.providerName, query)
-	return p.searchLdap(query, p.userScope, config, lConn)
+	return p.searchLdap(query, p.userScope, config, lConn, page, pageSize)
 }
 
-func (p *ldapProvider) searchGroup(name string, config *v3.LdapConfig, lConn ldapv3.Client) ([]v3.Principal, error) {
+func (p *ldapProvider) searchGroup(name string, config *v3.LdapConfig, lConn ldapv3.Client, page, pageSize int) ([]v3.Principal, error) {
 	if config.GroupSearchFilter != "" {
 		// Make sure group search filter contains a valid LDAP query expression
 		// before interpolating it into the search filter.
@@ -437,10 +437,10 @@ func (p *ldapProvider) searchGroup(name string, config *v3.LdapConfig, lConn lda
 	)
 
 	logrus.Debugf("%s searchGroup query: %s scope: %s", p.providerName, query, p.groupScope)
-	return p.searchLdap(query, p.groupScope, config, lConn)
+	return p.searchLdap(query, p.groupScope, config, lConn, page, pageSize)
 }
 
-func (p *ldapProvider) searchLdap(query string, scope string, config *v3.LdapConfig, lConn ldapv3.Client) ([]v3.Principal, error) {
+func (p *ldapProvider) searchLdap(query string, scope string, config *v3.LdapConfig, lConn ldapv3.Client, page, pageSize int) ([]v3.Principal, error) {
 	var principals []v3.Principal
 	var search *ldapv3.SearchRequest
 
@@ -470,12 +470,61 @@ func (p *ldapProvider) searchLdap(query string, scope string, config *v3.LdapCon
 		return nil, fmt.Errorf("ldap: error binding service account: %w", err)
 	}
 
-	results, err := lConn.SearchWithPaging(search, 1000)
-	if err != nil {
-		ldapErr, ok := err.(*ldapv3.Error)
-		if ok && ldapErr.ResultCode != ldapv3.LDAPResultNoSuchObject {
-			return []v3.Principal{}, fmt.Errorf("ldap: error searching for query %s:, error: %w", query, err)
+	var results *ldapv3.SearchResult
+	if pageSize > 0 {
+		if page < 1 {
+			page = 1
 		}
+		var err error
+		var pagingControl *ldapv3.ControlPaging
+		for i := 0; i < page; i++ {
+			if pagingControl == nil {
+				pagingControl = ldapv3.NewControlPaging(uint32(pageSize))
+				search.Controls = append(search.Controls, pagingControl)
+			} else {
+				for _, control := range search.Controls {
+					if c, ok := control.(*ldapv3.ControlPaging); ok {
+						c.SetCookie(pagingControl.Cookie)
+					}
+				}
+			}
+
+			results, err = lConn.Search(search)
+			if err != nil {
+				ldapErr, ok := err.(*ldapv3.Error)
+				if ok && ldapErr.ResultCode != ldapv3.LDAPResultNoSuchObject {
+					return []v3.Principal{}, fmt.Errorf("ldap: error searching for query %s:, error: %w", query, err)
+				}
+				break
+			}
+
+			pagingResult := ldapv3.FindControl(results.Controls, ldapv3.ControlTypePaging)
+			if pagingResult == nil {
+				if i < page-1 {
+					results.Entries = nil
+				}
+				break
+			}
+			pagingControl = pagingResult.(*ldapv3.ControlPaging)
+			if len(pagingControl.Cookie) == 0 {
+				if i < page-1 {
+					results.Entries = nil
+				}
+				break
+			}
+		}
+	} else {
+		results, err = lConn.SearchWithPaging(search, 1000)
+		if err != nil {
+			ldapErr, ok := err.(*ldapv3.Error)
+			if ok && ldapErr.ResultCode != ldapv3.LDAPResultNoSuchObject {
+				return []v3.Principal{}, fmt.Errorf("ldap: error searching for query %s:, error: %w", query, err)
+			}
+		}
+	}
+
+	if results == nil {
+		return principals, nil
 	}
 
 	for i := 0; i < len(results.Entries); i++ {
diff --git a/pkg/auth/providers/ldap/ldap_client_test.go b/pkg/auth/providers/ldap/ldap_client_test.go
index 66d92dfed..f19c0fbd7 100644
--- a/pkg/auth/providers/ldap/ldap_client_test.go
+++ b/pkg/auth/providers/ldap/ldap_client_test.go
@@ -596,7 +596,7 @@ func TestSearchLdapNoSuchObjectErrorIsIgnored(t *testing.T) {
 		},
 	}
 
-	principals, err := provider.searchLdap("(objectClass=inetOrgPerson)", provider.userScope, config, ldapConn)
+	principals, err := provider.searchLdap("(objectClass=inetOrgPerson)", provider.userScope, config, ldapConn, 0, 0)
 
 	require.NoError(t, err, "LDAPResultNoSuchObject should be treated as empty results, not an error")
 	require.Empty(t, principals)
@@ -626,8 +626,70 @@ func TestSearchLdapOtherLDAPErrorIsPropagated(t *testing.T) {
 		},
 	}
 
-	principals, err := provider.searchLdap("(objectClass=inetOrgPerson)", provider.userScope, config, ldapConn)
+	principals, err := provider.searchLdap("(objectClass=inetOrgPerson)", provider.userScope, config, ldapConn, 0, 0)
 
 	require.Error(t, err, "non-NoSuchObject LDAP errors should be propagated to the caller")
 	require.Empty(t, principals)
 }
+
+func TestSearchLdapWithPaging(t *testing.T) {
+	t.Parallel()
+
+	provider := &ldapProvider{
+		providerName: "openldap",
+		userScope:    "openldap_user",
+		groupScope:   "openldap_group",
+	}
+
+	config := &v3.LdapConfig{
+		LdapFields: v3.LdapFields{
+			ServiceAccountDistinguishedName: saDN,
+			ServiceAccountPassword:          saPassword,
+			UserObjectClass:                 userObjectClassName,
+			UserSearchBase:                  "ou=users,dc=foo,dc=bar",
+		},
+	}
+
+	callCount := 0
+	ldapConn := &ldapFakes.FakeLdapConn{
+		SearchFunc: func(searchRequest *ldapv3.SearchRequest) (*ldapv3.SearchResult, error) {
+			callCount++
+			if callCount == 1 {
+				return &ldapv3.SearchResult{
+					Entries: []*ldapv3.Entry{
+						{
+							DN: "cn=user1,ou=users,dc=foo,dc=bar",
+							Attributes: []*ldapv3.EntryAttribute{
+								{Name: ObjectClass, Values: []string{userObjectClassName}},
+							},
+						},
+					},
+					Controls: []ldapv3.Control{
+						ldapv3.NewControlPaging(1),
+					},
+				}, nil
+			}
+			return &ldapv3.SearchResult{
+				Entries: []*ldapv3.Entry{
+					{
+						DN: "cn=user2,ou=users,dc=foo,dc=bar",
+						Attributes: []*ldapv3.EntryAttribute{
+							{Name: ObjectClass, Values: []string{userObjectClassName}},
+						},
+					},
+				},
+				Controls: []ldapv3.Control{
+					// empty cookie
+				},
+			}, nil
+		},
+	}
+
+	// Request page 2, pageSize 1. It should make 2 calls.
+	principals, err := provider.searchLdap("(objectClass=inetOrgPerson)", provider.userScope, config, ldapConn, 2, 1)
+
+	require.NoError(t, err)
+	require.Len(t, principals, 1)
+	require.Equal(t, "openldap_user://cn=user2,ou=users,dc=foo,dc=bar", principals[0].ObjectMeta.Name)
+	require.Equal(t, 2, callCount)
+}
diff --git a/pkg/auth/providers/ldap/ldap_provider.go b/pkg/auth/providers/ldap/ldap_provider.go
index d2e87c964..15fa28133 100644
--- a/pkg/auth/providers/ldap/ldap_provider.go
+++ b/pkg/auth/providers/ldap/ldap_provider.go
@@ -184,7 +184,7 @@ func (p *ldapProvider) SearchPrincipals(searchKey, principalType string, myToken
 	}
 	defer lConn.Close()
 
-	principals, err = p.searchPrincipals(searchKey, principalType, config, lConn)
+	principals, err = p.searchPrincipals(searchKey, principalType, config, lConn, 0, 0)
 	if err == nil {
 		for _, principal := range principals {
 			switch principal.PrincipalType {
